src/card_operations/card_pair.c

```diff
@@ -0,0 +1,397 @@
+    if (false == pair_data->is_paired) {
+      card_data->nfc_data.status =
+          nfc_pair(pair_data->data, &(pair_data->data_len));
+
+      if (card_data->nfc_data.status != SW_NO_ERROR) {
+        card_handle_errors(card_data);
+      }
+    }
+  }
+  return;
+}
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+uint32_t card_pair_without_retap(uint8_t card_number) {
+  ASSERT(1 <= card_number && 4 >= card_number);
+
+  uint32_t error_status = DEFAULT_UINT32_IN_FLASH;
+  card_operation_data_t card_data = {
+      .error_message = NULL,
+      .error_type = CARD_OPERATION_DEFAULT_INVALID,
+      .nfc_data = {0}};
+  card_pairing_data_t pair_data = {0};
+
+  error_status = pair_card_preprocess(&pair_data);
+  if (SUCCESS != error_status) {
+    return error_status;
+  }
+
+  while (1) {
+    // Initialize card tap config
+    card_data.nfc_data.acceptable_cards = encode_card_number(card_number);
+
+    init_and_pair_card(&card_data, &pair_data);
+
+    if (CARD_OPERATION_SUCCESS == card_data.error_type) {
+      error_status = handle_pairing_success(&card_data, &pair_data);
+      buzzer_start(BUZZER_DURATION);
+      break;
+    }
+
+    if (CARD_OPERATION_CARD_REMOVED != card_data.error_type) {
+      break;
+    }
+  }
+
+  nfc_deselect_card();
+  return error_status;
+}
+
+uint32_t card_pair_operation(uint8_t card_number,
+                             char *heading,
+                             char *message) {
+  ASSERT(1 <= card_number && 4 >= card_number && NULL != message);
+
+  uint32_t error_status = DEFAULT_UINT32_IN_FLASH;
+  card_operation_data_t card_data = {
+      .error_message = NULL,
+      .error_type = CARD_OPERATION_DEFAULT_INVALID,
+      .nfc_data = {0}};
+  card_pairing_data_t pair_data = {0};
+
+  error_status = pair_card_preprocess(&pair_data);
+  if (SUCCESS != error_status) {
+    return error_status;
+  }
+
+  instruction_scr_init(message, heading);
+  card_data.nfc_data.retries = 5;
+
+  while (1) {
+    // Initialize card tap config
+    card_data.nfc_data.acceptable_cards = encode_card_number(card_number);
+
+    init_and_pair_card(&card_data, &pair_data);
+
+    if (CARD_OPERATION_SUCCESS == card_data.error_type) {
+      buzzer_start(BUZZER_DURATION);
+      error_status = handle_pairing_success(&card_data, &pair_data);
+      wait_for_card_removal();
+      break;
+    }
+
+    if ((CARD_OPERATION_CARD_REMOVED == card_data.error_type) ||
+        (CARD_OPERATION_RETAP_BY_USER_REQUIRED == card_data.error_type)) {
+      const char *error_msg = card_data.error_message;
+      if (CARD_OPERATION_SUCCESS == indicate_card_error(error_msg)) {
+        // Re-render the instruction screen
+        instruction_scr_init(message, heading);
+        continue;
+      }
+    } else {
+      error_status = card_data.nfc_data.status;
+      break;
+    }
```

Comment:
else clause is not required since `continue` is using in line #387

Reply:
updated